scripts/home_allocation.py

```diff
@@ -0,0 +1,138 @@
+sample_census_feature = nhs.data.sample_census_feature
+
+
+def main(
+    config_path: str,
+    gnaf_fname: str,
+    census_fname: str,
+    out_fname: str,
+    strategy: Literal["join_nearest", "filter"] | None,
+) -> None:
+    # Required for fiona - reads shapefiles
+    supported_drivers["ESRI Shapefile"] = "rw"
+
+    try:
+        data_config = nhs.config.data_config(config_path)
+        logger_config = nhs.config.logger_config(config_path)
+        simulation_config = nhs.config.simulation_config(config_path)
+    except Exception as e:
+        logger.critical(
+            f"Failed to load configuration at {config_path} with exception {e}, terminating..."
+        )
+        exit(1)
+
+    logger.enable("nhs")
+    nhs.logging.config_logger(logger_config)
+
+    gnaf_path = os.path.join(data_config["gnaf_path"], gnaf_fname)
+    logger.info(f"Reading GNAF data from {gnaf_path}...")
+    houses_df = read_parquet(gnaf_path)
+    if not isinstance(houses_df, pl.LazyFrame):
+        logger.error(f"Failed to load GNAF file at {gnaf_path}, terminating...")
+        exit(1)
+
+    census_path = os.path.join(data_config["census_path"], census_fname)
+    logger.info(f"Reading census data from {census_path}...")
+    census = read_parquet(census_path)
+    if not isinstance(census, pl.LazyFrame):
+        logger.error(f"Failed to load census file at {census_path}, terminating...")
+        exit(1)
+
+    logger.info(f"Reading shapefile from {data_config['shapefile_path']}...")
+    area_polygons: gpd.GeoDataFrame = read_shapefile(
+        data_config["shapefile_path"], data_config["crs"]
+    )
+
+    logger.info("Converting housing dataset to GeoDataFrame...")
+    house_coords = to_geo_dataframe(houses_df, data_config["crs"])
+
+    logger.info("Joining coordinates with area polygons...")
+    joined_coords = join_coords_with_area(house_coords, area_polygons, strategy)
+
+    out_path = os.path.join(data_config["output_path"], out_fname)
+    logger.info(
+        f"Randomly assigning census features to GNAF addresses and saving to {out_path}..."
+    )
+    census_gnaf = join_census_with_coords(census, joined_coords)
+
+    # TODO: is there a better way to handle column names instead of hard-coding?
+    allocated = randomly_assign_census_features(
+        census_gnaf,
+        "SA1_CODE_2021",
+        "LONGITUDE",
+        "LATITUDE",
+        simulation_config["census_features"],
+    )
+    logger.debug(allocated.explain(streaming=True))
+    allocated.collect().write_csv(out_path)
+    logger.info(f"Allocation complete, saved to {out_path}")
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Randomly assign census features to GNAF addresses."
+    )
+    parser.add_argument(
+        "-g",
+        "--gnaf_fname",
+        type=str,
+        help="Name of the input GNAF address geocode parquet file, e.g. 'TAS_ADDRESS_DEFAULT_GEOCODE_psv.parquet'",
+        required=True,
+    )
```

Comment:
This is another one that we will change to state as the input and have a dictionary mapping right, and that will be after we have it a one big parquet file.

Reply:
oh we will just change it to a **folder** instead of a single parquet file (will need to keep the comment about parquet files since that's the expected file format). We then get a dictionary of lazyframes, do some preprocessing and join them all together into one mega lazyframe (we are NOT writing the mega lazyframe to a parquet,, we might experience the heat death of the universe before it's done writing..)